plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java

```diff
@@ -163,12 +171,41 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
                 .putAdvancedOption(USER_AGENT_PREFIX, "")
                 .putAdvancedOption(USER_AGENT_SUFFIX, "Trino-exchange")
                 .build();
-        this.s3AsyncClient = createS3AsyncClient(
+        S3AsyncClient client = createS3AsyncClient(
                 credentialsProvider,
                 overrideConfig,
                 config.getAsyncClientConcurrency(),
                 config.getAsyncClientMaxPendingConnectionAcquires(),
                 config.getConnectionAcquisitionTimeout());
+        this.s3AsyncClient = new S3AsyncClientWrapper(client)
+        {
+            private final Map<RequestType, AtomicLong> requestsActive = new ConcurrentHashMap<>();
+            private final AtomicLong requestActiveLastLoggedTimestamp = new AtomicLong();
+            private final long logActiveRequestsIntervalInMillis = config.getLogActiveRequestsInterval().toMillis();
+
+            @Override
+            protected void handle(RequestType requestType, CompletableFuture<?> responseFuture)
+            {
+                requestsActive.computeIfAbsent(requestType, key -> new AtomicLong()).incrementAndGet();
+                logActiveRequestsIfNecessary();
+                responseFuture.whenComplete((result, failure) -> {
+                    requestsActive.get(requestType).decrementAndGet();
+                    logActiveRequestsIfNecessary();
+                    if (failure != null && failure.getMessage() != null && failure.getMessage().contains("Maximum pending connection acquisitions exceeded")) {
+                        log.error("Encountered the 'Maximum pending connection acquisitions exceeded' error. Requests active: %s", requestsActive);
+                    }
+                });
+            }
+
+            private void logActiveRequestsIfNecessary()
```

Comment:
Can’t this be exposed via JMX instead?

Reply:
Yeah, I think JMX is a much better option. I added simple logging to unblock debugging while working on JMX export. Implemented through JMX now.